src/types.lisp

```diff
@@ -569,18 +569,25 @@ The foreign array must be freed with foreign-array-free."
   ((count :initarg :count :accessor slot-count))
   (:documentation "Aggregate structure slots."))
 
-;;; Since MEM-REF returns a pointer for struct types we are able to
-;;; chain together slot names when accessing slot values in nested
-;;; structures.
 (defmethod foreign-struct-slot-value (ptr (slot aggregate-struct-slot))
-  "Return a pointer to SLOT relative to PTR."
-  (convert-from-foreign (inc-pointer ptr (slot-offset slot))
-                        (slot-type slot)))
+  "Return the value of an aggregate SLOT from a struct at PTR."
+  (let ((type (slot-type slot))
+        (input-ptr (inc-pointer ptr (slot-offset slot))))
+    (if (= 1 (slot-count slot))
+        (convert-from-foreign input-ptr type)
+        (loop for i below (slot-count slot)
+              collect (mem-aref input-ptr type i)))))
```

Comment:
I wonder if we should instead deprecate the `:count` option and convert `(x foo :count n)` slots to `(x (:array foo n))`. That way we wouldn't need specific code for handling `:count`.

Did you know you can do this?

```lisp
CFFI> (defcstruct test
        (x (:array :int 2)))
(:STRUCT TEST)
CFFI> (with-foreign-object (x '(:struct test))
        (setf (foreign-slot-value x '(:struct test) 'x) #(42 42))
        (foreign-slot-value x '(:struct test) 'x))
#(42 42)
```

(I didn't remember it was possible when I started reviewing your PR.)

Reply:
I didn't know about `:array`, but I'm not sure it works with C functions that take `(:struct test)`. If a C function returns a `(:struct test)` I will get, say, a list `(X #(1 2))`. If I then pass it to a C function, CFFI will convert the array to a foreign value and get a pointer (which is not the right thing to do), and then it will try to set the foreign slot's value to this pointer, but the slot is an `aggregate-struct-slot` so the value expected is a Lisp array.

I wondered how much code uses `:count` vs. `:array`. I scanned some of the Lisp libraries I have locally and came up with [this](https://gist.github.com/death/a7403e30d0b6eb3b759ad5901e745a0c). To generate this report I wrote [formgrep](https://github.com/death/formgrep). It's not surprising that `:array` is not used much, given that it's not documented, but maybe this should help a decision regarding deprecation of `:count`.

Reply:
The issue with `:array` you describe should be fixed as follows:

```lisp
(defmethod translate-into-foreign-memory ((object list)
                                          (type foreign-struct-type)
                                          p)
  (unless (bare-struct-type-p type)
    (loop for (name value) on object by #'cddr
          do (setf (foreign-struct-slot-value p (get-slot-info type name))
                   value))))
```

After that fix, the following snippet does what we expect, right?

```
CFFI> (defcstruct foo (x (:array :int 2)))
(:STRUCT FOO)
CFFI> (with-foreign-object (x '(:struct foo))
        (setf (mem-ref x '(:struct foo))
              (list 'x #(21 42)))
        (list (mem-aref x :int 0) (mem-aref x :int 1)))
(21 42)
```

Regarding :count vs :array, yeah, that's not unexpected. Should we go the deprecation route, we should deprecate in the documentation first, after a few years add a deprecating style warning, then a warning, then an error, then remove it. It shouldn't take more than 20 years. :D

Reply:
It seems there are still unexpected values; it probably needs some tests with actual C functions taking and returning such structures.